### ISSUE TITLE
[cfe] Avoid generating unnecessary type checking stubs for `covariant` parameters?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#46975]
Consider the following program:

```dart
class B {
  void m(num n) {}
}

abstract class I {
  void m(covariant num n);
}

class C extends B implements I {}

void main() {}
```

This class is specified to be a compile-time error (because `B.m` is inherited into a class where its parameter should be `covariant`), but the language team is working on a change that allows it, following the behavior that we have actually had in the CFE for years. So now the CFE doesn't have to change to make it an error, it's already doing the right thing.

However, the CFE is actually doing a little bit more than is necessary. Here is the kernel code obtained from compilation of the above program:

```
library /*isNonNullableByDefault*/;
import self as self;
import "dart:core" as core;

class B extends core::Object {
  synthetic constructor •() → self::B
    : super core::Object::•()
    ;
  method m(core::num n) → void {}
}
abstract class I extends core::Object {
  synthetic constructor •() → self::I
    : super core::Object::•()
    ;
  abstract method m(covariant core::num n) → void;
}
class C extends self::B implements self::I {
  synthetic constructor •() → self::C
    : super self::B::•()
    ;
  forwarding-stub method m(covariant core::num n) → void
    return super.{self::B::m}(n);
}
static method main() → void {}
```

Class `C` in the kernel code contains a declaration of the method `m` which is marked as a `forwarding-stub`, with a parameter that includes the modifier `covariant`.

I believe that this kind of forwarding stub need not be generated in this situation. In general, considering a single parameter:

Assume that `C` is a class that inherits a method `m` with a parameter `p` of type `T` which is not covariant-by-declaration as seen from the declaration of `m` which is inherited. Let _M<sub>I</sub>_ be the set of direct and indirect superinterfaces of `C`. Let _M<sub>p</sub>_ be the set of parameters in members of an element in _M<sub>I</sub>_ that correspond to `p`. A dynamic type check must be performed on the actual argument passed to `p` if there exists a parameter in _M<sub>p</sub>_ whose type is not a subtype of `T`.

In other words, we can omit the type checking stub in all cases where the inherited parameter has a type which is a supertype of all the other static types that we may ascribe to this parameter.

In the given example, all the parameters that correspond to each other have type `num`, and hence there is no way we could perform a statically checked invocation of `m` where we would pass an argument whose type isn't correct for the parameter of the method implementation that actually runs, and hence there's no need to have a forwarding stub.

@johnniwinther, @stefantsov, do you on the analysis of the situation? Do you think it will matter so much that it is useful to implement this optimization? If so, is there a backend part of this effort, or is it done entirely in the CFE?
